src/protocols/bdx/BDXMessageUtils.h

```diff
@@ -0,0 +1,204 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines structures and utility methods for working with BDX
+ *      messages.
+ */
+
+#pragma once
+
+#include <support/CodeUtils.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace chip {
+namespace BDX {
+
+enum ControlMode : uint8_t
+{
+    kNotSpecified  = 0x00,
+    kSenderDrive   = 0x10,
+    kReceiverDrive = 0x20,
+    kAsync         = 0x40,
+};
+
+/*
+ * A structure for representing a SendInit or ReceiveInit message (both contain
+ * identical parameters).
+ */
+struct TransferInit
+{
+public:
+    /**
+     * @brief
+     *  Pack (write) the message into a PacketBuffer.
+     *
+     * @param[out] aBuffer a PacketBuffer to use to write the message
+     *
+     * @return CHIP_ERROR Any error that occurs when trying to write to the PacketBuffer
+     */
+    CHIP_ERROR Pack(System::PacketBuffer * aBuffer);
+
+    /**
+     * @brief
+     *  Parse data from an PacketBuffer into a struct instance
+     *
+     * @param[in] aBuffer Pointer to a PacketBuffer containing the data
+     * @param[out] aParsedMessage Reference to a struct instance where results will be stored
```

Comment:
This does not make it at all clear that the resulting struct has pointers into the `PacketBuffer` and must not outlive the `PacketBuffer`, or any calls on the `PacketBuffer` that would rearrange its internal buffer in any way....

Is there a way we could make this safer?  Once #2707 is fixed, could we transfer ownership of the `PacketBuffer` to the struct, so no one else can mess with it and it will not get released until the struct is destroyed?

Similar for the other `Parse` methods here.

Reply:
In Weave, we had `RetainedPacketBuffer` which automatically reference-counted packetbuffers. The header still exists in `src/lib/protocols/common/CHIPMessage.h` but no implementation exists AFAIK. RetainedPacketBuffer is probably safest, but for now would a comment about ownership be enough? The only other alternative I see is using `MemoryAlloc()`or keeping a static buffer in the struct, but I'm not sure either of those are more desirable. What do you think?

Reply:
Again, if nothing except this struct will use the PacketBuffer, we can transfer ownership to it once #2707 is fixed; just file a followup issue tracking that.

If something else will use it, the struct can just take a separate reference to the PacketBuffer in addition to whatever the caller holds.  Again, followup is fine, but it would be best to not let it slide too long.

Reply:
Yea, this struct is just meant as a temporary organizational helper. It's not meant to exist outside of the context that it's called in, so it shouldn't outlive the packetbuffer. I'll add a comment explaining that, and open an issue to incorporate #2707 once it's fixed.